### PR TITLE
Adding support for creating and managing ACL Tokens 

### PR DESCRIPTION
**Pull request by armon**
This PR adds the concept of ACL Tokens. CRUD of the tokens are exposed via HTTP APIs, RPC endpoints, and the state store.

### CODE REVIEW

File: nomad/mock/mock.go [hunk 1]
```diff
@@ -357,3 +357,17 @@ func ACLPolicy() *structs.ACLPolicy {
 		ModifyIndex: 20,
 	}
 }
+
```
Comment (dadgar) [inline nomad/mock/mock.go]: May be nice to have another mock function to get a management token.

File: nomad/acl_endpoint.go [hunk 2]
```diff
@@ -148,3 +148,177 @@ func (a *ACL) GetPolicy(args *structs.ACLPolicySpecificRequest, reply *structs.S
...
+			out, err := state.ACLTokenByAccessorID(nil, token.AccessorID)
+			if err != nil {
+				return fmt.Errorf("token lookup failed: %v", err)
+			}
+			if out == nil {
+				return fmt.Errorf("cannot find token %s", token.AccessorID)
+			}
+
+			// Do not allow the secret ID or create time to be changed
+			token.SecretID = out.SecretID
```
Comment (dadgar) [inline nomad/acl_endpoint.go]: Why are you doing this here rather than in the state store?

Comment (armon) [inline nomad/acl_endpoint.go]: IDK does it matter?

File: nomad/state/state_store_test.go [hunk 3]
```diff
@@ -5881,6 +5881,189 @@ func TestStateStore_ACLPolicyByNamePrefix(t *testing.T) {
...
+		tk.AccessorID = prefix + tk.AccessorID[4:]
+		if err := state.UpsertACLTokens(baseIndex, []*structs.ACLToken{tk}); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+		baseIndex++
+	}
+
+	// Scan by prefix
+	iter, err := state.ACLTokenByAccessorIDPrefix(nil, "aa")
+	if err != nil {
```
Comment (chelseakomlo) [inline nomad/state/state_store_test.go]: assert.Nil(err)

File: nomad/state/state_store_test.go [hunk 4]
```diff
@@ -5881,6 +5881,189 @@ func TestStateStore_ACLPolicyByNamePrefix(t *testing.T) {
...
+	// Ensure we see both policies
+	count := 0
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		count++
+	}
+	if count != 2 {
```
Comment (chelseakomlo) [inline nomad/state/state_store_test.go]: assert.NotEqual(2, count)

**Pull request merged**